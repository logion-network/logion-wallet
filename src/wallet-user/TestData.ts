import { ValidAccountId } from "@logion/node-api";

export const TEST_WALLET_USER = ValidAccountId.polkadot("5H4MvAsobfZ6bBCDyj5dsrWYLrA8HrRzaqa9p61UXtxMhSCY");
export const TEST_WALLET_USER2 = ValidAccountId.polkadot("5EBxoSssqNo23FvsDeUxjyQScnfEiGxJaNwuwqBH2Twe35BX");
