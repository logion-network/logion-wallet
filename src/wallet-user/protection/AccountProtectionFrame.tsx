import { PendingRecovery, ProtectionParameters, ProtectionRequestStatus, RejectedRecovery } from "@logion/client";
import { Col, Row } from 'react-bootstrap';

import Frame from "src/common/Frame";
import { GREEN, ORANGE, YELLOW, rgbaToHex } from "src/common/ColorTheme";
import Icon from "src/common/Icon";
import { useLogionChain } from "src/logion-chain";
import ButtonGroup from "src/common/ButtonGroup";
import Button from "src/common/Button";
import { useCallback } from "react";
import { useUserContext } from "../UserContext";
import ExtrinsicSubmissionStateView from "src/ExtrinsicSubmissionStateView";
import RecoveryRefusal from "./RecoveryRefusal";
import SelectLegalOfficer from "./SelectLegalOfficer";
import { useCommonContext } from "src/common/CommonContext";

export type ProtectionRecoveryRequestStatus = 'pending' | 'accepted' | 'activated' | 'unavailable' | 'rejected';

export type Refusal = 'single' | 'double' | 'none';

export interface Props {
    type: ProtectionRecoveryRequestStatus;
}

export default function AccountProtectionFrame(props: Props) {
    const { submitCall, clearSubmissionState, extrinsicSubmissionState } = useLogionChain();
    const { availableLegalOfficers } = useCommonContext();
    const { protectionState, activateProtection, claimRecovery } = useUserContext();

    const activateProtectionCallback = useCallback(async () => {
        try {
            await submitCall(activateProtection!);
        } catch(e) {
            console.log(e);
        } finally {
            clearSubmissionState();
        }
    }, [ submitCall, activateProtection, clearSubmissionState ]);

    const claimRecoveryCallback = useCallback(async () => {
        try {
            await submitCall(claimRecovery!);
        } catch(e) {
            console.log(e);
        } finally {
            clearSubmissionState();
        }
    }, [ submitCall, claimRecovery, clearSubmissionState ]);

    if(protectionState === undefined || availableLegalOfficers === undefined) {
        return null;
    }

    const protectionParameters = protectionState.protectionParameters;
    const legalOfficer1 = protectionParameters.legalOfficers[0];
    const legalOfficer2 = protectionParameters.legalOfficers[1];
    let legalOfficer1Status: ProtectionRequestStatus;
    let legalOfficer2Status: ProtectionRequestStatus;
    if(protectionState instanceof PendingRecovery || protectionState instanceof RejectedRecovery) {
        legalOfficer1Status = protectionParameters.states[0].status;
        legalOfficer2Status = protectionParameters.states[1].status;
    } else if(props.type === 'accepted') {
        legalOfficer1Status = 'ACCEPTED';
        legalOfficer2Status = 'ACCEPTED';
    } else {
        legalOfficer1Status = 'ACTIVATED';
        legalOfficer2Status = 'ACTIVATED';
    }

    let refusal: Refusal = 'none';
    if (props.type === 'rejected') {
            refusal = protectionState.protectionParameters.states[1].status === 'REJECTED' ? 'double' : 'single';
    }

    return (
        <Frame className="ProtectionRecoveryRequest">
            <Header type={ props.type } protectionParameters={ protectionParameters } />

            {
                props.type === 'accepted' && !protectionParameters.isActive && !extrinsicSubmissionState.inProgress &&
                <ButtonGroup className="Activate">
                    <Button
                        data-testid="btnActivate"
                        onClick={ activateProtectionCallback }
                        variant="polkadot"
                    >
                        Activate
                    </Button>
                </ButtonGroup>
            }
            {
                props.type === 'activated' && protectionParameters.isRecovery && !protectionParameters.isClaimed && !extrinsicSubmissionState.inProgress &&
                <ButtonGroup className="Claim">
                    <Button
                        data-testid="btnClaim"
                        onClick={ claimRecoveryCallback }
                        variant="polkadot"
                    >
                        Claim
                    </Button>
                </ButtonGroup>
            }
            <ExtrinsicSubmissionStateView />

            { refusal === 'double' &&
                <Row className="legal-officers">
                    <Col md={ 6 }>
                    <RecoveryRefusal
                        recovery={ protectionState as RejectedRecovery }
                        refusal={ refusal }
                    />
                    </Col>
                </Row>
            }
            <Row className="legal-officers">
                <Col md={6}>
                    <SelectLegalOfficer
                        legalOfficerNumber={ 1 }
                        legalOfficer={ legalOfficer1 }
                        otherLegalOfficer={ legalOfficer2 }
                        legalOfficers={ availableLegalOfficers }
                        mode="view"
                        label="Legal Officer N°1"
                        status={ legalOfficer1Status }
                    />
                </Col>
                <Col md={6}>
                    { refusal === 'single' &&
                        <RecoveryRefusal
                            recovery={ protectionState as RejectedRecovery }
                            refusal={ refusal }
                        />
                    }
                    { refusal !== 'single' &&
                        <SelectLegalOfficer
                            legalOfficerNumber={ 2 }
                            legalOfficer={ legalOfficer2 }
                            otherLegalOfficer={ legalOfficer1 }
                            legalOfficers={ availableLegalOfficers }
                            mode="view"
                            label="Legal Officer N°2"
                            status={ legalOfficer2Status }
                        />
                    }
                </Col>
            </Row>

            <Row className="footer">
                <Col md={4} className="legal-officers-images">
                    <img className="legal-officer-male-image" alt="male legal officer" src={process.env.PUBLIC_URL + "/assets/landing/protection_male.svg"}/>
                    <img className="legal-officer-female-image" alt="female legal officer" src={process.env.PUBLIC_URL + "/assets/landing/protection_female.svg"}/>
                </Col>
                <Col md={8} className="legal-officers-text">
                    <p className="foundation"><strong>The foundation of</strong> your protection</p>
                    <p>
                        In charge of a <strong>public office</strong>, Logion Legal Officers are <strong>identified Judicial Officers</strong>,<br/>who apply a
                        strict code of ethics and are <strong>legaly responsible</strong> for their actions<br/>
                        while legally securing your digital assets and digital transactions.
                    </p>
                </Col>
            </Row>
        </Frame>
    );
}

interface HeaderProps {
    type: ProtectionRecoveryRequestStatus;
    protectionParameters: ProtectionParameters;
}

function Header(props: HeaderProps) {
    const forAccount = props.protectionParameters.isRecovery ? ` for account ${props.protectionParameters.recoveredAccount?.address}` : "";

    let icon;
    let color;
    let text;
    if(props.type === 'pending') {
        if(props.protectionParameters.isRecovery) {
            icon = "pending";
            color = ORANGE;
            text = `Your Logion Recovery request ${forAccount} has been submitted for review. Your Legal Officers will contact you
            as soon as possible to finalize the procedure.`;
        } else {
            icon = "pending";
            color = ORANGE;
            text = `Your Logion Protection request has been submitted for review. Please note that, after the successful
            completion of one of your Legal Officer approval processes, you will be able to use all features
            provided by your logion account dashboard.`;
        }
    } else if(props.type === 'accepted') {
        if(props.protectionParameters.isRecovery) {
            icon = "accepted";
            color = YELLOW;
            text = `Your recovery request has been accepted by your Legal Officers.
            You may now activate your protection. This will require 2 signatures.
            After that, you'll be able to initiate the actual recovery.`;
        } else {
            icon = "accepted";
            color = YELLOW;
            text = `Your Logion Protection request has been accepted by your
            Legal Officers. You may now activate your protection.`;
        }
    } else if(props.type === 'activated') {
        if(props.protectionParameters.isRecovery && !props.protectionParameters.isClaimed) {
            icon = "activated";
            color = GREEN;
            text = `You are now ready to claim the access to address ${props.protectionParameters.recoveredAccount?.address}.`;
        } else {
            icon = "activated";
            color = GREEN;
            text = `Your Logion Protection is active`;
        }
    } else if (props.type === 'rejected') {
        if(props.protectionParameters.isRecovery) {
            icon = "pending";
            color = ORANGE;
            text=`Your Logion Recovery request has been submitted. It appears that at least one of the
            selected Legal Officers refused your request: you have to decide what would be the next step.
            Please note that, only after the successful completion of both of your Legal Officer approval
            processes, you will be able to retrieve your assets from your previous account.`;
        } else {
            icon = "pending";
            color = ORANGE;
            text = `Your Logion Protection request has been submitted. It appears that at least one of the
            selected Legal Officers refused your request: you have to decide what would be the next step.
            Please note that, only after the successful completion of your Legal Officer approval processes,
            you will be able to use all features provided by your logion account dashboard.`;
        }
    } else {
        throw new Error("Unsupported type");
    }

    const backgroundColor = rgbaToHex(color, 0.2);
    return (
        <div
            className="header"
        >
            <Icon
                icon={{id: icon}}
            />
            <br/>
            <p
                className="text"
                style={{
                    color: color,
                    borderColor: color,
                    backgroundColor: backgroundColor,
                }}
            >
                { text }
            </p>
        </div>
    );
}
