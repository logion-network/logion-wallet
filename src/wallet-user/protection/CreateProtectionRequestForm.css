.CreateProtectionRequestFormInitiateRecovery {
    margin-bottom: 40px;
}

.CreateProtectionRequestFormInitiateRecovery .EstimatedFees {
    margin-bottom: 20px;
}

.CreateProtectionRequestFormInitiateRecovery .Button {
    margin-top: 20px;
}
