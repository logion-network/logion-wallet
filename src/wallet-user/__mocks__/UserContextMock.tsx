import { LogionNodeApiClass, Fees } from '@logion/node-api';

import { twoLegalOfficers } from "../../common/TestData";
import { TEST_WALLET_USER } from '../TestData';
import { AccountTokens, DirectoryClient, NoProtection, ProtectionState, LegalOfficerClass, DraftRequest } from '@logion/client';
import { AxiosFactory } from '@logion/client/dist/AxiosFactory.js';
import { ComponentFactory } from '@logion/client/dist/ComponentFactory.js';
import { LegalOfficerEndpoint, LogionClientConfig, SharedState } from '@logion/client/dist/SharedClient.js';
import { NetworkState } from '@logion/client/dist/NetworkState.js';
import { UserContext, ProtectionRequestParams } from '../UserContext';
import { BalanceState } from '@logion/client/dist/Balance.js';
import {
    LocsState,
    PendingRequest,
    ClosedLoc,
    RejectedRequest,
    ClosedCollectionLoc,
    OpenLoc, VoidedCollectionLoc, VoidedLoc
} from "@logion/client";
import { LocType, ValidAccountId } from "@logion/node-api";
import { Mock } from 'moq.ts';

export let createTokenRequest = () => null;

export let refreshRequests = jest.fn();

export let createProtectionRequest = jest.fn().mockReturnValue(Promise.resolve());

export function estimateFeesCreateProtectionRequest(_: ProtectionRequestParams): Promise<Fees> {
    return Promise.resolve(Fees.zero());
}

export let activateProtection = jest.fn().mockReturnValue(Promise.resolve());

export let claimRecovery = jest.fn().mockReturnValue(Promise.resolve());

export let api = new Mock<LogionNodeApiClass>();

export const DEFAULT_SHARED_STATE: SharedState = {
    axiosFactory: {} as AxiosFactory,
    componentFactory: {} as ComponentFactory,
    config: {

    } as LogionClientConfig,
    directoryClient: {} as DirectoryClient,
    legalOfficers: twoLegalOfficers,
    allLegalOfficers: twoLegalOfficers,
    networkState: {} as NetworkState<LegalOfficerEndpoint>,
    nodeApi: api.object(),
    tokens: {} as AccountTokens,
    currentAccount: {
        address: TEST_WALLET_USER,
        type: "Polkadot"
    } as unknown as ValidAccountId,
}

export let protectionState: ProtectionState | undefined = {} as NoProtection;

export function setProtectionState(state: ProtectionState | undefined) {
    protectionState = state;
}

export let recoveredBalanceState: BalanceState | undefined = {} as BalanceState;

export function setRecoveredBalanceState(state: BalanceState | undefined) {
    recoveredBalanceState = state;
}

export let mutateRecoveredBalanceState = jest.fn().mockReturnValue(Promise.resolve());

export let locsState: Partial<LocsState> | undefined = undefined;

export function setOpenedLocRequests(requests: any[]) {
    locsState = {
        get openLocs(): Record<LocType, OpenLoc[]> {
            return {
                "Transaction": requests,
                "Collection": [],
                "Identity": [],
            }
        }
    }
}

export function setClosedLocRequests(requests: any[]) {
    locsState = {
        get closedLocs(): Record<LocType, (ClosedLoc | ClosedCollectionLoc)[]> {
            return {
                "Transaction": requests,
                "Collection": [],
                "Identity": [],
            }
        }
    }
}

export function setVoidedLocs(requests: any[]) {
    locsState = {
        get voidedLocs(): Record<LocType, (VoidedLoc | VoidedCollectionLoc)[]> {
            return {
                "Transaction": requests,
                "Collection": [],
                "Identity": [],
            }
        }
    }
}

export function setPendingLocRequests(requests: any[]) {
    locsState = {
        get pendingRequests(): Record<LocType, PendingRequest[]> {
            return {
                "Transaction": requests,
                "Collection": [],
                "Identity": [],
            }
        }
    }
}

export function setRejectedLocRequests(requests: any[]) {
    locsState = {
        get rejectedRequests(): Record<LocType, RejectedRequest[]> {
            return {
                "Transaction": requests,
                "Collection": [],
                "Identity": [],
            }
        }
    }
}

export function setDraftLocRequests(requests: any[]) {
    locsState = {
        get draftRequests(): Record<LocType, DraftRequest[]> {
            return {
                "Transaction": requests,
                "Collection": [],
                "Identity": [],
            }
        }
    }
}

export function setOpenVerifiedIssuerLocs(transactions: any[], collections: any[]) {
    locsState = {
        get openVerifiedIssuerLocs(): Record<LocType, OpenLoc[]> {
            return {
                "Transaction": transactions,
                "Collection": collections,
                "Identity": [],
            }
        }
    }
}
export function setClosedVerifiedIssuerLocs(transactions: any[], collections: any[]) {
    locsState = {
        get closedVerifiedIssuerLocs(): Record<LocType, ClosedLoc[]> {
            return {
                "Transaction": transactions,
                "Collection": collections,
                "Identity": [],
            }
        }
    }
}

export function setHasNonVoidIdentityLoc(legalOfficers: LegalOfficerClass[]) {
    locsState = {
        legalOfficersWithNonVoidIdentityLoc: legalOfficers
    }
}

export function setLocsState(locsStateMock: LocsState) {
    locsState = locsStateMock;
}

export let mutateLocsState = jest.fn().mockReturnValue(Promise.resolve());

export function useUserContext() {
    return {
        refreshRequests,
        createProtectionRequest,
        estimateFeesCreateProtectionRequest,
        activateProtection,
        claimRecovery,
        protectionState,
        recoveredBalanceState,
        mutateRecoveredBalanceState,
        locsState,
        mutateLocsState,
    } as unknown as Partial<UserContext>;
}

export function setCreateTokenRequest(callback: any) {
    createTokenRequest = callback;
}

export function setRefreshRequests(func: any) {
    refreshRequests = func;
}

export function setCreateProtectionRequest(callback: any) {
    createProtectionRequest = callback;
}

export function setMutateLocsState(callback: any) {
    mutateLocsState = callback;
}

export function UserContextProvider() {
    return (
        <div/>
    );
}
