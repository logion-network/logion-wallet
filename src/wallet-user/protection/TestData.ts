import { DEFAULT_IDENTITY, DEFAULT_ADDRESS, DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER } from '../../common/TestData';
import { TEST_WALLET_USER } from '../TestData';
import { AccountRecoveryRequest, ProtectionRequestStatus } from '@logion/client/dist/AccountRecoveryClient';

export const IDENTITY_LOC_ID = "86f8884c-79fa-4d80-ba79-b7e3aa4728a7";

export const ACTIVATED_RECOVERY_REQUEST1: AccountRecoveryRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER.address, ANOTHER_LEGAL_OFFICER.address, "ACTIVATED", TEST_WALLET_USER.address, IDENTITY_LOC_ID);

export const ACTIVATED_RECOVERY_REQUEST2: AccountRecoveryRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER.address, DEFAULT_LEGAL_OFFICER.address, "ACTIVATED", TEST_WALLET_USER.address, IDENTITY_LOC_ID);

export const ACTIVATED_RECOVERY_REQUESTS = [ ACTIVATED_RECOVERY_REQUEST1, ACTIVATED_RECOVERY_REQUEST2 ];

export const PENDING_RECOVERY_REQUEST1: AccountRecoveryRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER.address, ANOTHER_LEGAL_OFFICER.address, "PENDING", TEST_WALLET_USER.address, IDENTITY_LOC_ID);

export const PENDING_RECOVERY_REQUEST2: AccountRecoveryRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER.address, DEFAULT_LEGAL_OFFICER.address, "PENDING", TEST_WALLET_USER.address, IDENTITY_LOC_ID);

export const PENDING_RECOVERY_REQUESTS = [ PENDING_RECOVERY_REQUEST1, PENDING_RECOVERY_REQUEST2 ];

export const RECOVERY_REQUEST: AccountRecoveryRequest = ACTIVATED_RECOVERY_REQUEST1;

export const ACCEPTED_RECOVERY_REQUEST1: AccountRecoveryRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER.address, ANOTHER_LEGAL_OFFICER.address, "ACCEPTED", TEST_WALLET_USER.address, IDENTITY_LOC_ID);

export const ACCEPTED_RECOVERY_REQUEST2: AccountRecoveryRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER.address, DEFAULT_LEGAL_OFFICER.address, "ACCEPTED", TEST_WALLET_USER.address, IDENTITY_LOC_ID);

export const ACCEPTED_RECOVERY_REQUESTS = [ ACCEPTED_RECOVERY_REQUEST1, ACCEPTED_RECOVERY_REQUEST2 ];

function createProtectionRequest(
    requestId: string,
    legalOfficerAddress: string,
    otherLegalOfficerAddress: string,
    status: ProtectionRequestStatus,
    addressToRecover: string,
    requesterIdentityLoc: string,
): AccountRecoveryRequest {
    return {
        id: requestId,
        requesterAddress: TEST_WALLET_USER.address,
        legalOfficerAddress,
        otherLegalOfficerAddress,
        decision: {
            rejectReason: status === 'REJECTED' ? "Because" : null,
            decisionOn: status !== 'PENDING' ? '2021-06-10T13:48:00.000Z' : null,
        },
        userIdentity: DEFAULT_IDENTITY,
        userPostalAddress: DEFAULT_ADDRESS,
        createdOn: '2021-06-10T13:48:00.000Z',
        addressToRecover,
        status,
        requesterIdentityLoc,
    };
}
