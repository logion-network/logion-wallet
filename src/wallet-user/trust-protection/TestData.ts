import { ProtectionRequest, ProtectionRequestStatus } from '@logion/client/dist/RecoveryClient.js';

import { DEFAULT_IDENTITY, DEFAULT_ADDRESS, DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER } from '../../common/TestData';
import { TEST_WALLET_USER } from '../TestData';

export const ACTIVATED_PROTECTION_REQUEST1: ProtectionRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "ACTIVATED", false, null);

export const ACTIVATED_PROTECTION_REQUEST2: ProtectionRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER, DEFAULT_LEGAL_OFFICER, "ACTIVATED", false, null);

export const ACTIVATED_PROTECTION_REQUESTS = [ ACTIVATED_PROTECTION_REQUEST1, ACTIVATED_PROTECTION_REQUEST2 ];

export const ACTIVATED_PROTECTION_REQUEST3: ProtectionRequest = createProtectionRequest("request3", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "ACTIVATED", false, null);

export const ACCEPTED_PROTECTION_REQUEST1: ProtectionRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "ACCEPTED", false, null);

export const ACCEPTED_PROTECTION_REQUEST2: ProtectionRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER, DEFAULT_LEGAL_OFFICER, "ACCEPTED", false, null);

export const ACCEPTED_PROTECTION_REQUESTS = [ ACCEPTED_PROTECTION_REQUEST1, ACCEPTED_PROTECTION_REQUEST2 ];

export const PENDING_PROTECTION_REQUEST1: ProtectionRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "PENDING", false, null);

export const PENDING_PROTECTION_REQUEST2: ProtectionRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER, DEFAULT_LEGAL_OFFICER, "PENDING", false, null);

export const PENDING_PROTECTION_REQUESTS = [ PENDING_PROTECTION_REQUEST1, PENDING_PROTECTION_REQUEST2 ];

export const PROTECTION_REQUEST: ProtectionRequest = ACTIVATED_PROTECTION_REQUEST1;

export const ACTIVATED_RECOVERY_REQUEST1: ProtectionRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "ACTIVATED", true, "toRecover");

export const ACTIVATED_RECOVERY_REQUEST2: ProtectionRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER, DEFAULT_LEGAL_OFFICER, "ACTIVATED", true, "toRecover");

export const ACTIVATED_RECOVERY_REQUESTS = [ ACTIVATED_RECOVERY_REQUEST1, ACTIVATED_RECOVERY_REQUEST2 ];

export const PENDING_RECOVERY_REQUEST1: ProtectionRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "PENDING", true, "toRecover");

export const PENDING_RECOVERY_REQUEST2: ProtectionRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER, DEFAULT_LEGAL_OFFICER, "PENDING", true, "toRecover");

export const PENDING_RECOVERY_REQUESTS = [ PENDING_RECOVERY_REQUEST1, PENDING_RECOVERY_REQUEST2 ];

export const RECOVERY_REQUEST: ProtectionRequest = ACTIVATED_RECOVERY_REQUEST1;

export const ACCEPTED_RECOVERY_REQUEST1: ProtectionRequest = createProtectionRequest("request1", DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, "ACCEPTED", true, "toRecover");

export const ACCEPTED_RECOVERY_REQUEST2: ProtectionRequest = createProtectionRequest("request2", ANOTHER_LEGAL_OFFICER, DEFAULT_LEGAL_OFFICER, "ACCEPTED", true, "toRecover");

export const ACCEPTED_RECOVERY_REQUESTS = [ ACCEPTED_RECOVERY_REQUEST1, ACCEPTED_RECOVERY_REQUEST2 ];

function createProtectionRequest(
    requestId: string,
    legalOfficerAddress: string,
    otherLegalOfficerAddress: string,
    status: ProtectionRequestStatus,
    isRecovery:boolean,
    addressToRecover:string | null
): ProtectionRequest {
    return {
        id: requestId,
        requesterAddress: TEST_WALLET_USER,
        legalOfficerAddress,
        otherLegalOfficerAddress,
        decision: {
            rejectReason: status === 'REJECTED' ? "Because" : null,
            decisionOn: status !== 'PENDING' ? '2021-06-10T13:48:00.000Z' : null,
        },
        userIdentity: DEFAULT_IDENTITY,
        userPostalAddress: DEFAULT_ADDRESS,
        createdOn: '2021-06-10T13:48:00.000Z',
        isRecovery,
        addressToRecover,
        status,
    };
}
