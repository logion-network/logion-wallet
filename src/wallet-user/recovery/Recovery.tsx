import { useUserContext } from "../UserContext";

import GoToTrustProtection from './GoToTrustProtection';
import CreateProtectionRequestForm from "../protection/CreateProtectionRequestForm";
import ProtectionRecoveryRequest from '../protection/ProtectionRecoveryRequest';
import RecoveryProcess from './RecoveryProcess';
import {
    AcceptedRecovery,
    NoProtection,
    PendingRecovery,
    UnavailableProtection,
    RejectedRecovery
} from "@logion/client";

export default function Recovery() {
    const { protectionState } = useUserContext();

    if (!protectionState || protectionState.discarded) {
        return null;
    }

    if(protectionState instanceof UnavailableProtection) {
        return <ProtectionRecoveryRequest type='unavailable' />;
    } else if(protectionState instanceof NoProtection) {
        return <CreateProtectionRequestForm isRecovery={ true } />;
    } else {
        const goToTrustProtection = !protectionState.protectionParameters.isRecovery && !protectionState.protectionParameters.isActive;
        if(goToTrustProtection) {
            return <GoToTrustProtection />;
        } else if(protectionState.protectionParameters.isActive) {
            if(protectionState.protectionParameters.isClaimed) {
                return <RecoveryProcess />;
            } else {
                return <ProtectionRecoveryRequest type='activated' />;
            }
        } else if(protectionState instanceof AcceptedRecovery) {
            return <ProtectionRecoveryRequest type='accepted' />;
        } else if(protectionState instanceof PendingRecovery) {
            return <ProtectionRecoveryRequest type='pending' />;
        } else if (protectionState instanceof RejectedRecovery) {
            return <ProtectionRecoveryRequest type='rejected' />;
        } else {
            return null;
        }
    }
}
