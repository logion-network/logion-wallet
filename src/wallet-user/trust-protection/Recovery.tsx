import { useUserContext } from "../UserContext";

import GoToTrustProtection from './GoToTrustProtection';
import CreateProtectionRequestForm from "./CreateProtectionRequestForm";
import ProtectionRecoveryRequest from './ProtectionRecoveryRequest';
import RecoveryProcess from './RecoveryProcess';
import {
    AcceptedProtection,
    NoProtection,
    PendingProtection,
    UnavailableProtection,
    RejectedRecovery
} from "@logion/client";

export default function Recovery() {
    const { protectionState } = useUserContext();

    if (!protectionState) {
        return null;
    }

    if(protectionState instanceof UnavailableProtection) {
        return <ProtectionRecoveryRequest type='unavailable' />;
    } else if(protectionState instanceof NoProtection) {
        return <CreateProtectionRequestForm isRecovery={ true } />;
    } else {
        const goToTrustProtection = !protectionState.protectionParameters.isRecovery && !protectionState.protectionParameters.isActive;
        if(goToTrustProtection) {
            return <GoToTrustProtection />;
        } else if(protectionState.protectionParameters.isActive) {
            if(protectionState.protectionParameters.isClaimed) {
                return <RecoveryProcess />;
            } else {
                return <ProtectionRecoveryRequest type='activated' />;
            }
        } else if(protectionState instanceof AcceptedProtection) {
            return <ProtectionRecoveryRequest type='accepted' />;
        } else if(protectionState instanceof PendingProtection) {
            return <ProtectionRecoveryRequest type='pending' />;
        } else if (protectionState instanceof RejectedRecovery) {
            return <ProtectionRecoveryRequest type='rejected' />;
        } else {
            return null;
        }
    }
}
