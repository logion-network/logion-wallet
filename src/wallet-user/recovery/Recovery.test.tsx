jest.mock('../UserContext');
jest.unmock("@logion/client");

import {shallowRender} from "../../tests";
import Recovery from "./Recovery";
import { setProtectionState } from '../__mocks__/UserContextMock';
import { PENDING_RECOVERY_REQUESTS, ACCEPTED_RECOVERY_REQUESTS, ACTIVATED_RECOVERY_REQUESTS } from "../protection/TestData";
import { PendingProtection, AcceptedProtection, PendingRecovery, ClaimedRecovery } from "@logion/client";
import { RecoverySharedState } from "@logion/client/dist/Recovery.js";
import { GUILLAUME, PATRICK } from "../../common/TestData";

test("renders", () => {
    setProtectionState(undefined);
    const tree = shallowRender(<Recovery />)
    expect(tree).toMatchSnapshot();
});

test("renders pending protection request", () => {
    setProtectionState(new PendingProtection({
        pendingProtectionRequests: PENDING_RECOVERY_REQUESTS,
        acceptedProtectionRequests: [],
        allRequests: PENDING_RECOVERY_REQUESTS,
        selectedLegalOfficers: [ PATRICK, GUILLAUME ]
    } as unknown as RecoverySharedState));

    const tree = shallowRender(<Recovery />)

    expect(tree).toMatchSnapshot();
});

test("renders accepted protection request", () => {
    setProtectionState(new AcceptedProtection({
        pendingProtectionRequests: [],
        acceptedProtectionRequests: ACCEPTED_RECOVERY_REQUESTS,
        allRequests: ACCEPTED_RECOVERY_REQUESTS,
        selectedLegalOfficers: [ PATRICK, GUILLAUME ]
    } as unknown as RecoverySharedState));

    const tree = shallowRender(<Recovery />)

    expect(tree).toMatchSnapshot();
});

test("renders protected but not claimed", () => {
    setProtectionState(new PendingRecovery({
        pendingProtectionRequests: [],
        acceptedProtectionRequests: ACCEPTED_RECOVERY_REQUESTS,
        allRequests: ACCEPTED_RECOVERY_REQUESTS,
        selectedLegalOfficers: [ PATRICK, GUILLAUME ],
        recoveryConfig: {
            legalOfficers: [ PATRICK.address, GUILLAUME.address ]
        }
    } as unknown as RecoverySharedState));

    const tree = shallowRender(<Recovery />)

    expect(tree).toMatchSnapshot();
});

test("renders protected and claimed", () => {
    setProtectionState(new ClaimedRecovery({
        pendingProtectionRequests: [],
        acceptedProtectionRequests: ACCEPTED_RECOVERY_REQUESTS,
        allRequests: ACCEPTED_RECOVERY_REQUESTS,
        selectedLegalOfficers: [ PATRICK, GUILLAUME ],
        recoveryConfig: {
            legalOfficers: [ PATRICK.address, GUILLAUME.address ]
        },
        recoveredAddress: ACTIVATED_RECOVERY_REQUESTS[0].addressToRecover!
    } as unknown as RecoverySharedState));

    const tree = shallowRender(<Recovery />)

    expect(tree).toMatchSnapshot();
});
