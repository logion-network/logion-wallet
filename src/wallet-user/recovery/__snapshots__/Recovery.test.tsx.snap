// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `null`;

exports[`renders accepted protection request 1`] = `
<ProtectionRecoveryRequest
  type="accepted"
/>
`;

exports[`renders pending protection request 1`] = `
<ProtectionRecoveryRequest
  type="pending"
/>
`;

exports[`renders protected and claimed 1`] = `<RecoveryProcess />`;

exports[`renders protected but not claimed 1`] = `
<ProtectionRecoveryRequest
  type="activated"
/>
`;
